metadata/src/utils/retain.rs

```diff
@@ -237,58 +165,191 @@ pub fn retain_metadata<F, G>(
         }
-        should_retain
-    });
+    }
 
-    // Additionally, subxt depends on the `DispatchError` type existing; we use the same
-    // logic here that is used when building our `Metadata`.
+    // 4. Collect all of the type IDs we still want to keep after deleting.
+    let mut keep_these_type_ids: BTreeSet<u32> = iterate_metadata_types(metadata)
+        .iter()
+        .map(|x| **x)
+        .collect();
+
+    // 5. Additionally, subxt depends on the `DispatchError` type existing; we use the same
+    // logic here that is used when building our `Metadata` to ensure we keep it too.
     let dispatch_error_ty = metadata
         .types
         .types
         .iter()
         .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
         .expect("Metadata must contain sp_runtime::DispatchError");
-    type_ids.insert(dispatch_error_ty.id);
 
-    // Now, keep the type IDs we've asked for. This recursively keeps any types referenced from these.
-    // This will return a map from old to new type ID, because IDs may change.
-    let map_ids = metadata.types.retain(|id| type_ids.contains(&id));
+    keep_these_type_ids.insert(dispatch_error_ty.id);
 
-    // And finally, we can go and update all of our type IDs in the metadata as a result of this:
-    update_outer_enums(&mut metadata.outer_enums, &map_ids);
-    for pallets in metadata.pallets.values_mut() {
-        update_pallet_types(pallets, &map_ids);
+    // 5. Strip all of the type IDs we no longer need, based on the above set.
+    let map_ids = metadata
+        .types
+        .retain(|id| keep_these_type_ids.contains(&id));
+
+    // 6. Now, update the type IDs referenced in our metadata to reflect this.
+    for id in iterate_metadata_types(metadata) {
+        if let Some(new_id) = map_ids.get(id) {
+            *id = *new_id;
+        } else {
+            panic!("Type id {id} was not retained. This is a bug");
+        }
     }
-    update_extrinsic_types(&mut metadata.extrinsic, &map_ids);
-    update_type(&mut metadata.runtime_ty, &map_ids);
-    update_runtime_api_types(metadata.apis.values_mut(), &map_ids);
+}
+
+fn strip_variants_in_enum_type<F>(metadata: &mut Metadata, mut pallets_filter: F, id: u32)
+where
+    F: FnMut(&str) -> bool,
+{
+    let ty = {
+        metadata
+            .types
+            .types
+            .get_mut(id as usize)
+            .expect("Metadata should contain enum type in registry")
+    };
+
+    let TypeDef::Variant(variant) = &mut ty.ty.type_def else {
+        panic!("Metadata type is expected to be a variant type");
+    };
+
+    variant.variants.retain(|v| pallets_filter(&v.name));
+}
+
+/// return an Vec handing back a mutable reference to each type ID seen in the metadata (not recursively).
+/// This will iterate over every type referenced in the metadata outside of `metadata.types`.
+fn iterate_metadata_types(metadata: &mut Metadata) -> alloc::vec::Vec<&mut u32> {
```

Comment:
This looks good for now, as actually iterating is much more ocmplicated and requires you to keep an internal state machine etc!

One thing I'd be tempted to do, since it's called `iterate_metadata_types` and only needed for iterating is to change the return type to `-> impl Iterator<Item=&mut u32>` and then call `types.into_iter()` at the end ofthe body to make that work.

(That way we can also optimise. the internals at some point without caring about anything else)

Reply:
done